### ISSUE TITLE
Missing fsync() call after linkat()

### ISSUE THREAD

**Issue by polarina** [andreacorbellini/rust-atomic-write-file#6]
https://github.com/andreacorbellini/rust-atomic-write-file/blob/1713ccce0e109f53094717528cc0f1afaa676e17/src/imp/unix/linux.rs#L47

A call to `fsync()` on a `O_TMPFILE` does nothing. Now, the anonymous file is given a temporary name then renamed to its final name, without an `fsync()` call in between. A power failure following the final rename may result in data loss -- the target file now contains no data, filesize is zero.

Take a look at the btrfs source code in the kernel, where the no-op behavior of `fsync()` is described: https://github.com/torvalds/linux/blob/026e680b0a08a62b1d948e5a8ca78700bfac0e6e/fs/btrfs/tree-log.c#L2422

**Comment by andreacorbellini** [comment]
Thanks for reporting this. I'm wondering if this more like a btrfs bug, rather than anything else: to start with, [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/functions/fsync.html) clearly states that data must be transferred to the storage device:

> The fsync() function shall request that all data for the open file descriptor named by fildes is to be transferred to the storage device associated with the file described by fildes.

It's clear why btrfs may want to skip flushing anonymous files. However, if the file is synced, linked, *and the directory is synced*, I'd expect the file to be written to the storage device, according to my interpretation of the standards. (Note that atomic-write-file does not fsync the directory: this is left as an optional operation to the caller.)

In any case, I've been trying to produce potential consistency issues with btrfs and ext4, with and without O_TMPFILE. So far I've not been successful, but to be fair I haven't been able to spend much time on this. Once I have some more time I'll try again.

**Comment by andreacorbellini** [comment]
I had a chance to take a closer look at this problem and do some experiments. Here are my findings.

**TL;DR:** there is a problem with using anonymous files on btrfs, which can leave the target files in an inconsistent state.

## Test methodology

A variety of strategies were tested, on both ext4 and btrfs:
* just plain `commit()`
* `commit()` followed by an `fsync()` on the file
* `commit()` followed by an `fsync()` on the directory
* `commit()` followed by an `fsync()` on the file and another `fsync()` on the directory (this was not necessary, and was included only for completeness)

For each of the above strategies, the following was performed:
1. create an empty filesystem
2. write some data to a test file (the "old content")
3. write some new data to the test file using `AtomicWriteFile` (the "new content")
4. run the `fsync()`s as mandated by the strategy
5. trigger a kernel panic
6. inspect the contents of the file

## Results

**with** `unnamed-tmpfile`:

|                           | ext4 | btrfs   |
|---------------------------|------|---------|
| commit                    | old  | old     |
| commit+sync dir           | new  | empty❗ |
| commit+sync file          | new  | new     |
| commit+sync file+sync dir | new  | new     |

**without** `unnamed-tmpfile`:

|                           | ext4 | btrfs |
|---------------------------|------|-------|
| commit                    | old  | old   |
| commit+sync dir           | new  | new   |
| commit+sync file          | new  | new   |
| commit+sync file+sync dir | new  | new   |

For the most part, these are the results I would expect. In particular, the `AtomicWriteFile` expectation is respected: the file is *either* in the old state, *or* in the new one, and *never* in an intermediate state.

The exception is of course on btrfs with unnamed-tmpfile with commit+sync dir: in that case the file is left in an intermediate state, which is not acceptable.

## Action Items

I still think this is a btrfs bug, which scans the journal from top to bottom ignoring anonymous files without considering that those files might be linked later in the journal, so I'll probably report it upstream. Nevertheless, this should probably be addressed in `AtomicWriteFile` for the time being.

I'll probably change the implementation to use `linkat()` followed by `fsync()` as suggested here, ~however note that this won't be a 100% failproof implementation, because `AtomicWriteFile` cannot control what happens between the `linkat()` and the consequent `fsync()`~ (I did not correctly remember the implementation: the solution will be failproof).

Another solution could be to avoid using anonymous temporary files on btrfs, but I'm not sure how easy and portable it is to do that, or if it would impact performance.

**Comment by andreacorbellini** [comment]
- Added an `fdatasync()` call after `linkat()` in f5799091589df847707efebc7f2324c993db6a50

  In a previous comment I said this would not be 100% failproof, but I was wrong: I did not remember my own code correctly. This implementation *should be* fully resistant to failures now.

- Added some automatic "crash tests" that simulate the behavior of the crate when kernel panics happen after a `commit()` (see the [README](https://github.com/andreacorbellini/rust-atomic-write-file/blob/master/crash-tests/linux/README.md) for details).

  Before the fix: [failure](https://github.com/andreacorbellini/rust-atomic-write-file/actions/runs/8823194185/job/24223112586)
  After the fix: [success](https://github.com/andreacorbellini/rust-atomic-write-file/actions/runs/8823989600/job/24225603338)

I'll make a release with the fix soon.

**Comment by andreacorbellini** [comment]
Fix released in [0.1.4](https://crates.io/crates/atomic-write-file/0.1.4)
